Check moab readability

jmartin-sul opened the pull request:
## Why was this change made?

To confirm that the contents of a Moab directory are readable before attempting to create a zip file from it, and to maybe kick Ceph MDS into making the files readable if MDS still sees the files as strays.

Per our hypothesizing in #1633, preservation_robots' implementation of similar logic (see https://github.com/sul-dlss/preservation_robots/pull/273) should (we hope) mitigate issues reading the new Moab version content.  But in the event there are still read issues, this will at least make it easier to retry the `zipmaker` failure without having to clean up a bad zip file from `/sdr-transfers`, by failing the job early (#1637 will also help keep retrying easy, if for some reason read issues crop up after the check added by this PR -- we haven't seen that in practice, but we're just getting a handle on the problem).

closes #1639 


## How was this change tested?

unit tests.  will also deploy to stage and run infra integration tests, and will confirm that content from tests was preserved.


## Which documentation and/or configurations were updated?

comments left where appropriate in code, w/ links to github issues.

ndushay commented on app/models/druid_version_zip.rb:
```diff
@@ -32,10 +32,11 @@ def base_key
     combined, status = Open3.capture2e(zip_command, chdir: work_dir.to_s)
     raise "zipmaker failure #{combined}" unless status.success?
-    raise "zip size (#{total_part_size}) is smaller than the moab size (#{moab_size})! zipmaker failure #{combined}" unless zip_size_ok?
+    raise "zip size (#{total_part_size}) is smaller than the moab size (#{moab_version_size})! zipmaker failure #{combined}" unless zip_size_ok?
```
love the method name change - how about adding version to the text here as well?

julianmorley replied:
Also maybe say that it the error message? "moab version size" instead of "moab size"?

jmartin-sul replied:
good catch, fixed!

jmartin-sul replied:
ah, yeah, @julianmorley that was what i interpreted @ndushay's suggestion to mean (sorry, GH didn't auto-refresh the page, and i didn't see this till i'd already pushed a change -- let me know if i missed something from either of your suggestions)

ndushay reviewed (approved):
one strong suggestion, o.w. LGTM

ndushay reviewed (changes requested):
actually, i will block the merge for my request ...

jcoyne commented on app/models/druid_version_zip.rb:
```diff
@@ -137,8 +140,18 @@ def zip_storage
 
   private
 
+  # should throw an error if any of the files in the moab are not yet readable, for example due to
```
```suggestion
  # Throws an error if any of the files in the moab are not yet readable.  For example due to
```

jcoyne commented on app/models/druid_version_zip.rb:
```diff
@@ -137,8 +140,18 @@ def zip_storage
 
+  # should throw an error if any of the files in the moab are not yet readable, for example due to
+  # Ceph MDS instance for a pres cat worker VM thinking that a file is a stray as a result of our
+  # particular use of hardlinking in preservation ingest.  allows for a quick directory walk before
```
```suggestion
  # particular use of hardlinking in preservation ingest.  Allows for a quick directory walk before
```

jcoyne commented on app/models/druid_version_zip.rb:
```diff
@@ -137,8 +140,18 @@ def zip_storage
+  # should throw an error if any of the files in the moab are not yet readable, for example due to
+  # Ceph MDS instance for a pres cat worker VM thinking that a file is a stray as a result of our
+  # particular use of hardlinking in preservation ingest.  allows for a quick directory walk before
+  # attempting to create the zip file(s).  see https://github.com/sul-dlss/preservation_catalog/issues/1633
```
```suggestion
  # attempting to create the zip file(s).  See https://github.com/sul-dlss/preservation_catalog/issues/1633
```

jcoyne commented on app/models/druid_version_zip.rb:
```diff
@@ -137,8 +140,18 @@ def zip_storage
+  # particular use of hardlinking in preservation ingest.  allows for a quick directory walk before
+  # attempting to create the zip file(s).  see https://github.com/sul-dlss/preservation_catalog/issues/1633
+  # @raise [StandardError] if storage_location is not available (should have been provided in constructor)
+  # @raise [Errno::EACCES, ?] if it is not possible to stat one or more files in the Moab
```
```suggestion
  # @raise [Errno::EACCES] if it is not possible to stat one or more files in the Moab
```

jmartin-sul replied:
So in this case, I was not entirely sure what the full list of plausible errors might be.  Admittedly, I didn't look back through all of the HB alerts when I first did this, but I wanted to call out that any number of ruby file read errors could be raised.

Looking back through Honeybadger alerts now, I also see some others that've occurred when reading Moab contents, which I'll added to the list, resulting in:  `Errno::EACCES, Errno::EIO, Errno::ENOENT, Errno::ESTALE`.

Still, looking over all the possibilities for error codes in `Errno.constants`, it's clear that most are not relevant to this method, but I suspect that some which we haven't seen are also plausible error codes.  For example, `EBUSY`, `ENOMEM`, and `EPERM`.

My bad for not being more thorough in listing the errors we've already encountered, but I would like to indicate that there are errors which might be raised, but which we're unsure about.  Though if you know for sure that the error codes I mentioned speculatively can't come up, happy to leave off the `?`.

i realize that the `?` may not technically be valid YARD syntax, but to my knowledge we don't auto-generate documentation off the YARD comments, and i though there was more value in expressing the uncertainty about possible errors than there was in producing strictly compliant YARD comments.

(I also just spent about 10 minutes, which was 10 minutes more than i should've, reading a 2003 Slashdot thread about some controversy with the `EDOOFUS` error code.  Same as it ever was 😆 )

jmartin-sul replied:
feel free to file an issue if you want me to circle back on this one again.

jcoyne commented on app/models/druid_version_zip.rb:
```diff
@@ -147,12 +160,14 @@ def total_part_size
 
-  def moab_size
+  def moab_version_size
+    moab_version_files.map { |file_path| File.size(file_path) }.sum
```
```suggestion
    moab_version_files.sum { |file_path| File.size(file_path) }
```

jmartin-sul replied:
good catch, and fixed on the prior method definition too

jcoyne commented on spec/models/druid_version_zip_spec.rb:
```diff
@@ -3,6 +3,9 @@
 
 describe DruidVersionZip do
+  # some tests below will use the version of the constructor that takes storage_location, because they
+  # are testing behavior where a zip is being created from a moab on disk.  for tests that don't exercise
```
```suggestion
  # are testing behavior where a zip is being created from a Moab on disk.  For tests that don't exercise
```

jcoyne commented on spec/models/druid_version_zip_spec.rb:
```diff
@@ -3,6 +3,9 @@
 require 'rails_helper'
 
 describe DruidVersionZip do
+  # some tests below will use the version of the constructor that takes storage_location, because they
```
```suggestion
   # Some tests below will use the constructor that takes storage_location, because they
```

jcoyne commented on spec/models/druid_version_zip_spec.rb:
```diff
@@ -97,6 +100,26 @@
       end
+
+      it 'calls check_moab_version_readability! before attempting to create the zip' do
+        # normally checking call order is to be avoided, but we want to make sure that we only attempt zip
+        # creation after proactively checking to see that all files to be zipped are readable.  readability
+        # checking logic is tested elsewhere in this spec.
```
```suggestion
        # This checks that we only attempt zip creation after checking to see that all files are readable.
```

jcoyne commented on spec/models/druid_version_zip_spec.rb:
```diff
@@ -295,4 +318,35 @@
+      ]
+    end
+
+    # note: if this test is dropped or altered, revisit the #create_zip! tests in this spec w.r.t.
```
Why would this test be dropped or altered?

jmartin-sul replied:
not totally sure, and i was a bit on the fence about including this comment (and the corresponding one on line 105).

i don't think this test is strictly necessary in terms of code coverage results, and so i can imagine it getting deleted accidentally, or intentionally as cruft/overkill, and then not getting noticed if the PR review is cursory.  i wanted to avoid that.  it's also testing some behavior that seems unnecessary at first glance, so it was easy to imagine aggressive code pruning removing it.

if that scenario seems unlikely to you, and you think this adds too much noise, i can remove the comment.

jcoyne replied:
I think this is confusing. I think this comment could apply to almost every test case so I'm not clear why we need to call this out in particular.

jmartin-sul replied:
fair enough, got rid of it

julianmorley commented on app/models/druid_version_zip.rb:
```diff
@@ -137,8 +140,18 @@ def zip_storage
+
   def zip_size_ok?
-    total_part_size >= moab_size
+    total_part_size >= moab_version_size
```
Really should be `>` here; the zip overhead dictates that the sum of all parts *must* be larger than moab_version_size on disk.

jmartin-sul replied:
good call, done

jmartin-sul commented:
@jcoyne and @julianmorley, i think i've addressed your comments, mind giving another look?

@jcoyne i implemented yours on my local instance and squashed them into the existing commits just to keep from having a messy history or squashing everything, but thanks for using suggestions, i do usually like that mechanism.  also there was one change that i didn't implement exactly as suggested, but happy to revisit if you think it needs more fixing.

jmartin-sul commented:
@jcoyne i _think_ i've addressed all of your feedback.  since we saw a small batch of `zipmaker` errors yesterday that this would likely help catch just a bit earlier, i wanted to get this merged and deployed this morning.  apologies if this was presumptuous.
